R/SEIRAge.R

```diff
@@ -0,0 +1,275 @@
+    # also raise errors if initial state and parameter values are not doubles
+    for (p in list('S0', 'E0', 'I0', 'R0')){
+      if(length(ic[[p]]) != object@n_age_categories){
+        stop(glue('Wrong number of age groups for {p}
+              compartments.'))}
+      if(!is.numeric(ic[[p]])){
+        stop(glue('{p} format must be numeric'))}
+    }
+    if(sum(S0, E0, I0, R0) != 1){
+      stop('All compartments need to sum up to 1.')
+    }
+    
+    # if all above tests are passed, assign the ic namelist to the object
+    object@initial_conditions <- ic
+    
+    return(object)
+  })
+
+# Setter and getter methods for transmission_parameters of an age-structured
+# SEIR model.
+
+#' @describeIn SEIRAge Retrieves transmission_parameters for an
+#' age-structured SEIR model.
+#'
+#' @param object An object of the class SEIRAge.
+#' 
+#' @return Transmission parameters of SEIRAge model.
+setGeneric('transmission_parameters',
+           function(object) standardGeneric('transmission_parameters'))
+
+setMethod('transmission_parameters', 'SEIRAge',
+          function(object) object@transmission_parameters)
+
+#' @describeIn  SEIRAge 
+#' @description Sets transmission_parameters of an
+#' age-structured SEIR model.
+#' 
+#' If the transmission parameters provided to are not 1-dimensional an error is
+#' thrown.
+#' 
+#' @param b rate at which an infected individual exposes susceptible.
+#' @param k rate at which exposed individuals become infected.
+#' @param g rate at which infected individuals recover.
+#'
+#' All rates of change between compartments are equal regardless of
+#' age group.
+#'
+#' @return Updated version of the age-structured SEIR model.
+setGeneric(
+  'transmission_parameters<-',
+  function(object, b, k, g){
+    standardGeneric('transmission_parameters<-')
+  })
+
+setMethod(
+  'transmission_parameters<-', 'SEIRAge',
+  function(object, b, k, g) {
+
+    # create list of parameter values
+    trans_params <- list(b, k, g)
+
+    # add names to each value
+    names(trans_params) = object@transmission_parameter_names
+
+    # check format of parameters b, k and g
+    if(length(b) != 1 | length(k) != 1 | length(g) != 1){
+      stop('The parameter values should be 1-dimensional.')
+    }
+
+    # if all above tests are passed, assign the trans_params namelist to the
+    # object
+    object@transmission_parameters <- trans_params
+
+    return(object)
+  })
+
+# Method to simulate output using from SEIRAge model.
+
+#' @describeIn SEIRAge Solves a system to ODEs which form an
+#' age-structured simple SEIR model. The system of equations for the time
+#' evolution of population fractions in Susceptible (S), Exposed (E), Infected
+#' (I) and Recovered (R) groups in a given age group indexed by i is given by
+#'
+#' \deqn{\frac{dS_i(t)}{dt} = - b S_i(t) I_i(t)}
+#' \deqn{\frac{dE_i(t)}{dt} =  b S_i(t) I_i(t) - k E_i(t)}
+#' \deqn{\frac{dI_i(t)}{dt} = k E_i(t) - g I_i(t)}
+#' \deqn{\frac{dR_i(t)}{dt} = g I_i(t)}
```

Comment:
Not sure if it's a common thing to do in R, but maybe we can give some reference for the model?

Reply:
Yep, good idea.